### DISCUSSION TITLE
Checking for RTMP:// or srt picture bitrate

### DISCUSSION THREAD

**Discussion by ghost** [WarmUpTill/SceneSwitcher#860] [General]
"Hello, Did anyone was able to do a check on the bitrate for Stream? I know I can do a macro base on the video output / true or false. However, I was thinking more on something when the bitrate dropped but I am still connected.

Based on 3 senses: good, low, drop 100% no signal.

I can do that with other software, etc., but it would be nice to have all in one place."

thanks all !

**Comment by ghost** [comment]
Digging more I can probably do it with the source gopro settings match...

I think I can build up on there.

**Comment by WarmUpTill** [comment]
Just to clarify:
Are you referring to the overall bitrate of the stream or the bitrate of a particular source type?

In case of the former you can do something like this:

![image](https://github.com/WarmUpTill/SceneSwitcher/assets/19472752/77da0e10-ece4-40c9-9548-b4696982a0d9)

Not sure about the latter.
In that case can you give more details how you usually check the bitrate of a given source?

**Comment by ghost** [comment]
Ty for the tip.
The idea is to switch scenes when the video GoPro" media source" loses connection or has a low connection.
This idea is the same as loopy_srt_monitor or Noalbs

**Normal Transmission will have:**
Scene - Live
GoPro - Media source "RTMP/SRT"
I will say anything above 2500


**Low connection Scenes will have:**
GoPro - Media source "RTMP/SRT"
Text (GDI+) - " low connection hold on"
Anything lower than 1800

**offline connection will have**
GoPro - Media source "RTMP/SRT"
Text (GDI+) - " OPS. WE LOST THE STREAMER LOOKING FOR SIGNAL"
This we lost the signal totally.

Thanks you guys :)

**Comment by ghost** [comment]
I found this online can be used as a guide as well.
![image](https://github.com/WarmUpTill/SceneSwitcher/assets/25457328/6a341394-0669-418d-b509-6d01900295f5)

**Comment by WarmUpTill** [comment]
I see - so this is not information that is available directly to OBS itself and must be pulled from _somewhere_?
Or am I misunderstanding how this works?

**Comment by dgraid** [comment]
Hello! Have the same question. I want to switch scene if rtmp signal is lost, and switch back again when it has signal.
 
Trying to get Media State on Media Source, but there is nothing similar there.

<img width="991" alt="CleanShot 2023-09-11 at 17 08 55@2x" src="https://github.com/WarmUpTill/SceneSwitcher/assets/44290678/2de42151-bf5f-4710-a929-99a4840032ae">

**Comment by WarmUpTill** [comment]
I had a look at the source code of the aforementioned [NOLABS](https://github.com/NOALBS/nginx-obs-automatic-low-bitrate-switching).
Is my understanding correct that it simply [checks](https://github.com/NOALBS/nginx-obs-automatic-low-bitrate-switching/blob/6560daee67d789da222e0940e751c886f990f2fd/src/stream_servers/obs.rs#L61-L74) if the media source is in state `playing` for three seconds for the stream to be online and if not switches to the configured offline scene?

If so then that is already possible using a macro setup similar to this:
![image](https://github.com/WarmUpTill/SceneSwitcher/assets/19472752/5a1fb4da-fbb5-458a-8d7f-ce3f74670838)
![image](https://github.com/WarmUpTill/SceneSwitcher/assets/19472752/43b7818d-e923-4ea5-aa35-b53416adcf40)

Please do let me know if that understanding is wrong! :)

For the other streaming services / protocols it seems to query a given URL for the stats of the stream and performs the scene switch based on the returned information.
That _might_ also be possible using the Http action / File condition.

**Reply by dgraid** [reply]
Hello, thank you for the answer. Yes you understand our goal right. And your method seams to solve it, but it does not work.

The macro does not see the difference in state of media. 

https://share.cleanshot.com/wMTwZZJH 
Look here, i made some video with test. And macro reacts on ALL states of the Media! I thing something is wrong here, check it yourself.

How it should work:
If no connection to the rtmp stream, then state of Media should be "buffering" or "stoped" or "error" or at least "Not playing", so we could switch to other scene, and when it is "playing" than we could switch back again.

**Comment by ghost** [comment]
I have as the state has = ended. 

I think the idea is to check back to Live if Gopro "RTMP" is back then Live should be the main again. 

![image](https://github.com/WarmUpTill/SceneSwitcher/assets/25457328/9eed5059-5aa9-4e65-ad69-ce098dc56e0a)

**Comment by WarmUpTill** [comment]
I experimented a bit with a local rtmp server and made the following observations:

* If a connection to the given address is established successfully the media source is considered in state "playing" by OBS.

![Playing](https://github.com/WarmUpTill/SceneSwitcher/assets/19472752/21493260-dded-4352-8376-851504c80f30)

* The media source is also in state "playing" for a brief amount of time, while it is attempting to reconnect to the given address.
So it is important to consider the value configured for this option, as even a disconnected media source will be in the state "playing" every X seconds.

![Reconnect](https://github.com/WarmUpTill/SceneSwitcher/assets/19472752/b870183e-bfd8-454d-a561-98b0b6836c50)

* A media source which has lost its connection to the rtmp server is considered to be in state "Ended" by OBS.
With the exception of the reconnect behaviour mentioned above.

![Ended](https://github.com/WarmUpTill/SceneSwitcher/assets/19472752/9d1438fb-ef47-4e58-8b49-cb659ca865bc)

* OBS will *not* automatically restart the connection if the source is not visible.
So if the media source is not part of all scenes which you want to have this automation work on, it cannot function.
Alternatively you can open a windowed projector of the media source to ensure that it is always active.

So the macro setup suggested originally should work as expected *under the assumption* that the configured [duration modifier](https://github.com/WarmUpTill/SceneSwitcher/wiki/Macro-duration-modifiers) is at least larger than the amount of time it takes for OBS to check if a reconnect is possible. (In my case OBS attempted this for about 5 seconds before giving up regardless of the configured reconnect delay)

Here is an example of the setup in action:
![Example](https://github.com/WarmUpTill/SceneSwitcher/assets/19472752/13fff4f1-9d56-4d6d-be57-be2922d44069)

If you want to [import](https://github.com/WarmUpTill/SceneSwitcher/wiki/Exporting-and-importing-individual-macros) the exact macros I used in the example you can use this code:
```
{"macros":[{"name":"Live","pause":false,"parallel":false,"onChange":true,"skipExecOnStart":false,"group":false,"dockSettings":{"register":false,"hasRunButton":true,"hasPauseButton":true,"hasStatusLabel":false,"highlightIfConditionsTrue":false,"runButtonText":"Run","pauseButtonText":"Pause","unpauseButtonText":"Unpause","conditionsTrueStatusText":"Conditions are true.","conditionsFalseStatusText":"Conditions are false."},"macroActionConditionSplitterPosition":[{"pos":160},{"pos":223}],"macroElseActionSplitterPosition":[{"pos":185},{"pos":0}],"registerHotkeys":false,"pauseHotkey":[],"unpauseHotkey":[],"togglePauseHotkey":[],"conditions":[{"collapsed":false,"id":"media","logic":0,"durationModifier":{"time_constraint":1,"seconds":{"value":{"value":10.0,"type":0},"unit":0,"version":1}},"source":{"type":0,"name":"Media Source"},"sceneSelection":{"type":0,"name":""},"sourceType":0,"state":1,"restriction":0,"duration":{"value":{"value":0.0,"type":0},"unit":0,"version":1},"matchOnChagne":false,"version":0}],"actions":[{"collapsed":false,"id":"scene_switch","enabled":true,"sceneSelection":{"type":0,"name":"live"},"transitionType":1,"duration":{"value":{"value":0.0,"type":0},"unit":0,"version":1},"blockUntilTransitionDone":true,"sceneType":0}],"elseActions":[]},{"name":"Offline","pause":false,"parallel":false,"onChange":true,"skipExecOnStart":false,"group":false,"dockSettings":{"register":false,"hasRunButton":true,"hasPauseButton":true,"hasStatusLabel":false,"highlightIfConditionsTrue":false,"runButtonText":"Run","pauseButtonText":"Pause","unpauseButtonText":"Unpause","conditionsTrueStatusText":"Conditions are true.","conditionsFalseStatusText":"Conditions are false."},"macroActionConditionSplitterPosition":[{"pos":192},{"pos":191}],"macroElseActionSplitterPosition":[{"pos":153},{"pos":0}],"registerHotkeys":false,"pauseHotkey":[],"unpauseHotkey":[],"togglePauseHotkey":[],"conditions":[{"collapsed":false,"id":"macro","logic":1,"durationModifier":{"time_constraint":0,"seconds":{"value":{"value":0.0,"type":0},"unit":0,"version":1}},"macros":[],"macro":"Live","type":1,"condition":0,"count":{"value":0,"type":0},"multiStateCount":{"value":0,"type":0},"multiStateCondition":2,"actionIndex":{"value":1,"type":0},"version":1}],"actions":[{"collapsed":false,"id":"scene_switch","enabled":true,"sceneSelection":{"type":0,"name":"offline"},"transitionType":0,"transition":"","duration":{"value":{"value":0.0,"type":0},"unit":0,"version":1},"blockUntilTransitionDone":true,"sceneType":0}],"elseActions":[]}],"version":"a166813"}
```
But note that you might have to tweak the exact values depending on how your OBS behaves in regards to reconnection to the server.
And the source and scene names will have to be adjusted as well.

Let me know if you have any questions! :)

---

P.S.: @dgraid as _none_ of the media states matched in your example video I think your issue is probably caused by something else.
Can you maybe open a separate issue and share a log file?

**Reply by dgraid** [reply]
Cool! I will check all this, and will come with feedback! Thank you!!!
